Microsoft.Toolkit/Diagnostics/Generated/Guard.Collection.tt

```diff
@@ -0,0 +1,259 @@
+        /// Asserts that the input <#=item.XmlType#> instance must be empty.
+        /// </summary>
+        /// <typeparam name="T">The item of items in the input <#=item.XmlType#> instance.</typeparam>
+        /// <param name="<#=item.Name#>">The input <#=item.XmlType#> instance to check the size for.</param>
+        /// <param name="name">The name of the input parameter being tested.</param>
+        /// <exception cref="ArgumentException">Thrown if the size of <paramref name="<#=item.Name#>"/> is != 0.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsEmpty<T>(<#=item.Type#> <#=item.Name#>, string name)
+        {
+            if (<#=item.Name#>.<#=item.Size#> != 0)
+            {
+                ThrowHelper.ThrowArgumentExceptionForIsEmpty(<#=item.Cast#><#=item.Name#>, name);
+            }
+        }
+
+        /// <summary>
+        /// Asserts that the input <#=item.XmlType#> instance must not be empty.
+        /// </summary>
+        /// <typeparam name="T">The item of items in the input <#=item.XmlType#> instance.</typeparam>
+        /// <param name="<#=item.Name#>">The input <#=item.XmlType#> instance to check the size for.</param>
+        /// <param name="name">The name of the input parameter being tested.</param>
+        /// <exception cref="ArgumentException">Thrown if the size of <paramref name="<#=item.Name#>"/> is == 0.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsNotEmpty<T>(<#=item.Type#> <#=item.Name#>, string name)
+        {
+            if (<#=item.Name#>.<#=item.Size#> == 0)
+            {
+<#
+    if (item.Type == "Span<T>")
+    {
+#>
+                ThrowHelper.ThrowArgumentExceptionForIsNotEmptyWithSpan<T>(name);
+<#
+    }
+    else if (item.Type == "ReadOnlySpan<T>")
+    {
+#>
+                ThrowHelper.ThrowArgumentExceptionForIsNotEmptyWithReadOnlySpan<T>(name);
+<#
+    }
+    else
+    {
+#>
+                ThrowHelper.ThrowArgumentExceptionForIsNotEmpty<<#=item.Type#>>(name);
+<#
+    }
+#>
+            }
+        }
+
+        /// <summary>
+        /// Asserts that the input <#=item.XmlType#> instance must have a size of a specified value.
+        /// </summary>
+        /// <typeparam name="T">The item of items in the input <#=item.XmlType#> instance.</typeparam>
+        /// <param name="<#=item.Name#>">The input <#=item.XmlType#> instance to check the size for.</param>
+        /// <param name="size">The target size to test.</param>
+        /// <param name="name">The name of the input parameter being tested.</param>
+        /// <exception cref="ArgumentException">Thrown if the size of <paramref name="<#=item.Name#>"/> is != <paramref name="size"/>.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void HasSizeEqualTo<T>(<#=item.Type#> <#=item.Name#>, int size, string name)
+        {
+            if (<#=item.Name#>.<#=item.Size#> != size)
```

Comment:
Are we worried about null case for all these?

Reply:
Same as comment above ([link](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131#discussion_r390925589)), the same concept applies there. All the collections are not marked as nullable, to the IDE will show a warning for you. Plus this way we get the `Guard` APIs to only perform the same exact checks they declare in the name. If you call an API called `IsNotNull`, it will check for `null`. If you call an API called `HasSizeEqualTo`, it will do just that. This makes the whole thing as straightforward as possible as well, other than also reducing the overhead 😊